### ISSUE TITLE
Thread leak, continuously growing threadpool with splitio initialized

### ISSUE THREAD

**Issue by kyleadams-fm** [splitio/dotnet-client#40]
For the last several weeks we noticed that the number of active threads in the thread pool will continuously grow when we have splitio initialized. This is most interestingly happening when our system is idle and not serving any traffic. The problem is that when we start to serve traffic and our number of active threads is larger than the minthreads setting, we are seeing long queues pile up for serving requests. 
This seems to indicate that something in splitIO is creating threads and never properly releasing them. 
You can see this happening in the image below over the course of 24 hours without serving any other traffic
![image](https://user-images.githubusercontent.com/73129760/122302164-9428be80-ceb6-11eb-99de-025bfa1a93c9.png)

If left on long enough without a restart we have even seen this number of threads grow to almost 500 active threads
![image](https://user-images.githubusercontent.com/73129760/122302694-50828480-ceb7-11eb-8872-094bda2d8987.png)

When we first noticed this we were on the deprecated .net-core-client package so we upgraded to this version and also noticed that there was a fix recently for a memory leak https://github.com/splitio/dotnet-client/pull/37 that we thought might help alleviate this issue but neither have fixed this issue of thread growth.

**Comment by alexanderscott** [comment]
@sanzmauro any thoughts on what could be causing threads to not be properly disposed and returned to the threadpool for reuse stemming from this Split package?

**Comment by chillaq** [comment]
Hi @kyleadams-fm , this is typically a symptom of using multiple SDK factories, each time the factory is initialized, it will create multiple threads to sync the SDK cache in the background, these threads do not stop even if the factory is not used.
Could you verify if this is the case and make sure the factory is created as singleton only? You can re-use the factory everytime you need to calculate treatments.

Thanks
Bilal

**Comment by kyleadams-fm** [comment]
@chillaq thanks for the response! 
We do create the object as a Singleton and use that throughout. Here is a sample of the code we use to initialize and create a singleton.
```
public static void AddSplitIo(this IServiceCollection services, SplitIoInfrastructureSettings settings)
        {
            if (!settings.UseMock)
            {
                var featureFlagService = new FmFeatureFlags(
                    splitIoApiKey: settings.SplitIoApiKey,
                    config: settings.SplitIoConfigurationOptions ?? new ConfigurationOptions());

                featureFlagService.Initialize().Wait();

                services.AddSingleton<IFeatureFlags>(featureFlagService);
            }
            else
            {
                services.AddSingleton<IFeatureFlags, NoopFeatureFlags>();
            }
        }
```

**Comment by alexanderscott** [comment]
@chillaq worth noting as well that we see this growth of active threads when the service is completely idle.  meaning no API handlers or background tasks or anything is actually calling into Split library code to accidentally cause client reinstantiation.

this thread problem should be reproducible by running a basic .NET server (we're using .NET 5) with Split client initialized in a singleton with default Split configuration and routinely dumping threadpool stats (we're use prometheus to collect this data).  it's only apparent after several hours, but should see average number of active threads in the threadpool grow linearly over time.  I tried this locally on my laptop the other day and saw the same effect.

**Comment by chillaq** [comment]
Thanks @alexanderscott , could you please provide the basic code using the SDK you used to repro the issue?

**Comment by kyleadams-fm** [comment]
@chillaq Here is the singleton we're creating to interact with the sdk to get treatments

```
namespace Fm.Utils.Features
{
    public class ClientNotInitializedError : Exception
    {
    }

    public class FmFeatureFlags : IFeatureFlags
    {
        public bool IsInitialized { get; private set; }
        private const string SplitOn = "on";
        private const int SplitInitializationWait = 10000;
        private readonly ISplitClient _sdk;

        public FmFeatureFlags(string splitIoApiKey, ConfigurationOptions config)
        {
            var factory = new SplitFactory(apiKey: splitIoApiKey, options: config);
            _sdk = factory.Client();
        }

        public Task Initialize()
        {
            return Task.Run(
                    () => _sdk.BlockUntilReady(SplitInitializationWait))
                .ContinueWith((antecedent) => IsInitialized = antecedent.IsCompletedSuccessfully);
        }

        public bool IsEnabled(string feature, string key, Dictionary<string, string>? context = null)
        {
            return IsOnFromSplitSdk(feature, key, context);
        }

        public string GetTreatment(string feature, string key, Dictionary<string, string>? context = null)
        {
            return GetSplitSdkTreatment(feature, key, context);
        }

        private string GetSplitSdkTreatment(string feature, string key, Dictionary<string, string>? context)
        {
            if (!IsInitialized)
            {
                throw new ClientNotInitializedError();
            }

            string treatment;

            if (context != null)
            {
                var splitResult = _sdk.GetTreatmentWithConfig(key, feature,
                    new Dictionary<string, object>(MakeDictionaryGeneric(context!)));
                treatment = splitResult.Treatment;
            }
            else
            {
                treatment = _sdk.GetTreatment(key, feature);
            }

            return treatment;
        }

        private bool IsOnFromSplitSdk(string feature, string key, Dictionary<string, string>? context)
        {
            return GetSplitSdkTreatment(feature, key, context) == SplitOn;
        }

        private Dictionary<string, object> MakeDictionaryGeneric(Dictionary<string, string> dict)
        {
            return dict.ToDictionary<KeyValuePair<string, string>, string, object>(
                entry => entry.Key,
                entry => entry.Value);
        }
    }
}

```

**Comment by sanzmauro** [comment]
Hi @kyleadams-fm @alexanderscott,

I'm ping you just to tell you that I'm working on this and I will back to you when I fix the issue. 
Thanks for sharing this information. 

Regards,
Mauro

**Comment by alexanderscott** [comment]
@chillaq @sanzmauro thank you for looking into this.  I pushed up a barebones sample repo which demonstrates the issue here https://github.com/flymachine-dev/dotnet-grpc-splitio-sample.  Hope it helps.

**Comment by alexanderscott** [comment]
hi @sanzmauro, any update on this?  we would like to continue using Split server-side, but this thread bug is affecting runtime performance

**Comment by sanzmauro** [comment]
hi @alexanderscott, i released a candidate release [6.3.2-rc339](https://www.nuget.org/packages/Splitio/6.3.2-rc339) and i am doing testing to be confident that we fixed the issue. 

You can use this rc if you want. I will ping you when i release the stable version if all is good. 

Regards

**Comment by alexanderscott** [comment]
thanks!  the [diff](https://github.com/splitio/dotnet-client/compare/threads-leak-fix) looks good to me 👍 .  I'd be happy to help test it against our [example repo](https://github.com/flymachine-dev/dotnet-grpc-splitio-sample) sometime later this week to verify

**Comment by sanzmauro** [comment]
Excellent, please let me know if all is ok from your side after test it 💪

**Comment by alexanderscott** [comment]
@sanzmauro seeing a significant improvement from your fix on [6.3.2-rc339](https://www.nuget.org/packages/Splitio/6.3.2-rc339).  

captured thread stats using `6.3.2-rc339` Split package before and after 72hrs idle in the sample repo with Split client running.  in that time, process thread count only went from 54 to 67 - which is a big improvement 😄 .  full prometheus report outputs here: https://github.com/flymachine-dev/dotnet-grpc-splitio-sample/compare/6.3.2-rc339

please let us know once you have finalized the RC and released a new version so we can point our services to it.  thanks again

**Comment by sanzmauro** [comment]
Hi @alexanderscott,

I released 6.3.2 with the fix.

I will close the issue, but feel free to re-open it if you need.

Regards
Mauro
